doc/neps/nep-0018-array-function-protocol.rst

```diff
@@ -438,25 +497,50 @@ make it impossible to implement this generic fallback behavior for
 ``__array_function__``.
 
 We could resolve this issue by change the handling of return values in
-``__array_function__`` in either of two possible ways: 1. Change the
-meaning of all arguments returning ``NotImplemented`` to indicate that
-all arguments should be coerced to NumPy arrays instead. However, many
-array libraries (e.g., scipy.sparse) really don't want implicit
-conversions to NumPy arrays, and often avoid implementing ``__array__``
-for exactly this reason. Implicit conversions can result in silent bugs
-and performance degradation. 2. Use another sentinel value of some sort
-to indicate that a class implementing part of the higher level array API
-is coercible as a fallback, e.g., a return value of
-``np.NotImplementedButCoercible`` from ``__array_function__``.
-
-If we take this second approach, we would need to define additional
-rules for how coercible array arguments are coerced, e.g., - Would we
-try for ``__array_function__`` overloads again after coercing coercible
-arguments? - If so, would we coerce coercible arguments one-at-a-time,
-or all-at-once?
-
-These are slightly tricky design questions, so for now we propose to
-defer this issue. We can always implement
+``__array_function__`` in either of two possible ways:
+
+1. Change the meaning of all arguments returning ``NotImplemented`` to indicate
+   that all arguments should be coerced to NumPy arrays instead. However, many
+   array libraries (e.g., scipy.sparse) really don't want implicit conversions
+   to NumPy arrays, and often avoid implementing ``__array__`` for exactly this
+   reason. Implicit conversions can result in silent bugs and performance
+   degradation.
+2. Use another sentinel value of some sort, e.g.,
+   ``np.NotImplementedButCoercible``, to indicate that a class implementing part
+   of NumPy's higher level array API is coercible as a fallback. This is a more
+   appealing option.
+
+If we take this second approach, we would need to define additional rules for
+how coercible array arguments are coerced. For sanity, a return value of
+``np.NotImplementedButCoercible`` from ``__array_function__`` should be handled
+equivalently to the corresponding argument not defining an
+``__array_function__`` method at all. If all arguments to a NumPy function
+either have no ``__array_function__`` method defined or return
+``NotImplementedButCoercible`` from ``__array_function__``, we would fall back
+to NumPy's current behavior: all arguments would be coerced to NumPy arrays.
+
+This approach also suggests an intriguing possibility: default implementations
+of ``__array_function__`` and ``__array_ufunc__`` on ``numpy.ndarray`` could
+change to simply return ``NotImplementedButCoercible``, i.e.,
```

Comment:
I would find it extremely unexpected behaviour for `ndarray` itself to return this: the semantics of `__array_function__` is that it provides an answer if it can, and in this case `ndarray` of course can provide an answer.  To me, this strongly suggests this would be a design mistake.

Instead, please do discuss the alternative of `ndarray.__array_function__` simply calling the unwrapped function - that would also remove the need to treat its `__array_function__` differently, and has the benefit of actually doing what it is supposed to do.